src/lib/input/input-wrapper.ts

```diff
@@ -0,0 +1,289 @@
+  private set _inputType(value: string) {
+    this._getterSetterOnlyInputType = value || 'text';
+    this._validateInputType();
+  }
+  private _getterSetterOnlyInputType = 'text';
+
+  /** The value of the `input` or `textarea`. */
+  private _inputValue = '';
+
+  /** The native `input` or `textarea` element. */
+  private _inputElement: HTMLInputElement | HTMLTextAreaElement;
+
+  /** A `MutationObserver` to observe the `input` or `textarea`. */
+  private _inputObserver = new MutationObserver(mutations => {
+    for (let mutation of mutations) {
+      switch (mutation.attributeName) {
+        case 'disabled':
+          this._inputDisabled = this._inputElement.disabled;
+          break;
+        case 'id':
+          this._inputId = this._inputElement.id;
+          break;
+        case 'type':
+          // We need to use getAttribute since `type="date"` is not supported on Firefox and returns
+          // `"text"` if we just do `inputElement.type`.
+          this._inputType = this._inputElement.getAttribute('type');
+          break;
+        case 'placeholder':
+          this._inputPlaceholder = this._inputElement.placeholder;
+          break;
+        case 'required':
+          this._inputRequired = this._inputElement.required;
+          break;
+      }
+    }
+  });
+
+  /** A map of event listeners to install on the `input` or `textarea`. */
+  private _inputListeners: { [key: string]: () => void } = {
+    'blur': () => { this._focused = false; },
+    'focus': () => { this._focused = true; },
+    'input': () => { this._inputValue = this._inputElement.value; }
+  };
+
+  constructor(private _elementRef: ElementRef) {}
+
+  ngAfterContentInit() {
+    this._initInputEl();
+    this._validateHints();
+
+    // Trigger validation when the hint children change.
+    this._hintChildren.changes.subscribe(() => {
+      this._validateHints();
+    });
+  }
+
+  ngOnChanges(changes: {[key: string]: SimpleChange}) {
+    if (changes['hintLabel']) {
+      this._validateHints();
+    }
+  }
+
+  /** Set focus on the input element. */
+  _focusInput() {
+    if (this._inputElement) {
+      this._inputElement.focus();
+    }
+  }
+
+  /** Whether the input has a placeholder. */
+  _hasPlaceholder(): boolean {
+    return !!this._inputPlaceholder || !!this._placeholderChild;
+  }
+
+  /** Initialize the `input` or `textarea` element. */
+  private _initInputEl() {
+    // Find input or textarea element.
+    let inputEls = this._elementRef.nativeElement.querySelectorAll('input, textarea');
+    if (inputEls.length != 1) {
+      new MdInputWrapperInputElementCountError(inputEls.length);
+    }
+    this._inputElement = inputEls[0];
+
+    // Install mutation observer and event listeners and subscribe to ngModel changes.
+    this._inputObserver.observe(this._inputElement, {
+      attributes: true,
+      attributeFilter: ['disabled', 'id', 'type', 'placeholder', 'required']
+    });
+    for (let eventType in this._inputListeners) {
```

Comment:
I think you don't need this extra handling around `this._inputListeners`; it's less code overall (and less indirection) to just do
```ts
inputElement.addEventListener('blur',  () => { this._focused = false; });
inputElement.addEventListener('focus', () => { this._focused = true; });
inputElement.addEventListener('input', () => { this._inputValue = this._inputElement.value; });
```